### PR TITLE
Blocks and panels

### PR DESCRIPTION
**Pull request by nassimdonald**
#### What does this PR do? (please provide any background)

* Added new 'block' classes to provide basic spacing for content – see [Blocks](http://localhost:3000/molecules/blocks/index.html) in Pattern Library for more info.
* Added new background classes and background-variant mixin to provide a quick way of styling content with different colour palettes – see [Blocks](http://localhost:3000/molecules/blocks/index.html) in Pattern Library for more info.
* Extensive improvements to panels and panel-groups, providing a more opinionated format for their use in layouts and alongside aforementioned background variations.

**Note:** Once tried and tested, I propose that this work is propagated to a global level, where it can be utilised by all brands consistently. There is still substantial work to be done on the spacing of components at a more general level, so some of this work is likely to change as we approach the problem more globally – but it's a start at the very least. 😃

#### What tests does this PR have?
None.

#### How can this be tested?

1. Checkout the branch `blocks-and-panels` in both the Brands repo and the Pattern Library repo.
2. In the directory of Pattern Library, run `npm link {path-to-your-brands-repo}`.
3. In Pattern Library, run `npm run build` followed by `npm run start`.
4. Visit the [Blocks](http://localhost:3000/molecules/blocks/index.html) and [Panels](http://localhost:3000/molecules/panels/index.html) pages in Pattern Library.

#### Screenshots / Screencast

pattern library - blocks
pattern library - panels

#### What gif best describes how you feel about this work?

---

- [x] I have checked our general [contributing document](https://github.com/holidayextras/culture/blob/master/CONTRIBUTING.md) and the project specific [contributing document](../blob/master/CONTRIBUTING.md) (if present) and I'm happy for this to be reviewed.

#### Reviewers

**Review 1**
- [ ] :+1:

**Review 2**
- [ ] :+1:

**Review 3**
- [ ] :+1:

By adding a +1 you are confirming you have...
- Witnessed the work behaving as expected (this could be on the author's machine or screencast).
- Checked for coding anti-patterns.
- Checked for appropriate test coverage.
- Checked all the tests are passing.

### CODE REVIEW

File: brands/holidayextras/src/less/_blocks.less [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//
+// Blocks
+// --------------------------------------------------
+
+
+// Base classes
+// --------------------------------------------------
+// Sizing
+.block-xs {
```
Comment (jodiedoubleday) [inline brands/holidayextras/src/less/_blocks.less]: I think maybe this should be moved to globals?

File: brands/holidayextras/src/less/_list_group.less [hunk 2]
```diff
@@ -54,6 +55,12 @@
 .list-group-item-variant(danger; @state-danger-bg; @state-danger-text; @brand-danger-light);
 
 
+// List group item badges
+// -------------------------
+.list-group-item .badge {
+  margin-top: floor((@line-height-computed - @font-size-small - 6px)/2);
```
Comment (jodiedoubleday) [inline brands/holidayextras/src/less/_list_group.less]: This seems very complex, is the 6px here half the `@font-size-small`

File: brands/holidayextras/src/less/_variables.less [hunk 3]
```diff
@@ -54,9 +56,10 @@
 @headings-font-family: "Yanone Kaffeesatz", "Helvetica Neue", Helvetica, Arial, sans-serif;
 @headings-font-weight: 400;
 @headings-line-height: 1;
+@headings-color-inverse: @white;
 
 //** Unit-less `line-height` for use in components like buttons.
-@line-height-base: 1.61803398875; // Golden Ratio
+@line-height-base: 1.5; // Golden Ratio
```
Comment (jodiedoubleday) [inline brands/holidayextras/src/less/_variables.less]: If not golden ratio, remove comment

File: brands/holidayextras/src/less/_variables.less [hunk 4]
```diff
@@ -80,15 +83,18 @@
 @border-radius-large:      10px;
 @border-radius-small:      2px;
 
-
+@block-padding-xs:         @padding-base-vertical;
```
Comment (jodiedoubleday) [inline brands/holidayextras/src/less/_variables.less]: Again maybe globals?

**Comment by stuartgiles** [comment]
Taking 2️⃣

File: brands/holidayextras/src/less/_blocks.less (hunk 1 above)
Comment (stuartgiles) [inline brands/holidayextras/src/less/_blocks.less]: Agreed, along with the default variable settings too. Great for theming later.

File: brands/holidayextras/src/less/_blocks.less [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  }
+  // Temporary fix for page-header spacing
+  .page-header:first-child {
+    margin-top: 0;
+  }
+}
+
+// Make a block pop!
+.block-pop {
```
Comment (stuartgiles) [inline brands/holidayextras/src/less/_blocks.less]: Should `pop` be associated with `blocks`?
For me, `blocks` feel like they should be reserved for just spacing.
Maybe it should be part of a wider set of elevation or shadow effects e.g. `elevation-xs`, `elevation-sm`, `elevation-md` and `elevation-lg`. Like [Material Design Elevation](https://material.google.com/what-is-material/elevation-shadows.html) these classes would be designed to reinforce information heir-achy and spacial relationships between elements?

File: brands/holidayextras/src/less/_backgrounds.less [hunk 6]
```diff
@@ -0,0 +1,34 @@
+//
+// Contextual backgrounds
+// --------------------------------------------------
+
+.bg-inverse {
+  background-color: @header-blue-dark;
+  .background-variant-inverse(@bg: @header-blue);
+}
+.bg-normal {
```
Comment (stuartgiles) [inline brands/holidayextras/src/less/_backgrounds.less]: Sorry I'm a bit confused about `bg-normal` and when I use it. Not necessarily a problem.

File: brands/holidayextras/src/less/_list_group.less (hunk 2 above)
Comment (nassimdonald) [inline brands/holidayextras/src/less/_list_group.less]: It's not ideal, but it is a calculation based on the properties of Bootstrap's `.badge` component – which has 6px (not a variable 😢) total vertical padding plus a line-height of `@font-size-small * @badge-line-height`. Actually, I should probably include `@badge-line-height` in the calculation too then.

File: brands/holidayextras/src/less/_variables.less (hunk 3 above)
Comment (nassimdonald) [inline brands/holidayextras/src/less/_variables.less]: Well spotted!

File: brands/holidayextras/src/less/_blocks.less (hunk 1 above)
Comment (nassimdonald) [inline brands/holidayextras/src/less/_blocks.less]: Moving to globals if all in agreement.

File: brands/holidayextras/src/less/_blocks.less (hunk 5 above)
Comment (nassimdonald) [inline brands/holidayextras/src/less/_blocks.less]: Sure, sounds good. It was just a quick tongue-in-cheek thing anyway, I think your suggestion makes more sense in practice.
For the record, not a big fan of how Material Design looks with regards to shadow effects – but that's personal preference. I feel there are more elegant ways to design in the z-axis, without the overuse of cheap looking box-shadows – look at iOS for example.

File: brands/holidayextras/src/less/_backgrounds.less (hunk 6 above)
Comment (nassimdonald) [inline brands/holidayextras/src/less/_backgrounds.less]: This is to address the issue of nesting. For example, if you have a panel with `bg-default` as a child of a block with `bg-inverse`, then you use `bg-normal` on the panel to "reset" (un-invert) the colours. It's not ideal, but with the nature of CSS, I can't see any other way.

**Comment by nassimdonald** [comment]
This PR is now invalid, as *blocks*, *panels* and *backgrounds* work are being divided into separate PRs.

**Pull request closed without merging**